master/internal/rm/kubernetesrm/informer.go

```diff
@@ -43,38 +46,108 @@ func newInformer(
 		cb:         cb,
+		name:       name,
 		syslog:     syslog,
 		resultChan: rw.ResultChan(),
 	}, nil
 }
 
-// startInformer returns the updated pod, if any.
-func (i *informer) run() {
-	i.syslog.Debug("pod informer is starting")
-	for event := range i.resultChan {
-		if event.Type == watch.Error {
-			i.syslog.Warnf("pod informer emitted error %+v", event)
-			continue
-		}
+func newEventInformer(
+	ctx context.Context,
+	eventInterface typedV1.EventInterface,
+	namespace string,
+	cb informerCallback,
+) (*informer, error) {
+	events, err := eventInterface.List(ctx, metaV1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
 
-		pod, ok := event.Object.(*k8sV1.Pod)
-		if !ok {
-			i.syslog.Warnf("error converting event of type %T to *k8sV1.Pod: %+v", event, event)
-			continue
-		}
+	rw, err := watchtools.NewRetryWatcher(events.ResourceVersion, &cache.ListWatch{
+		WatchFunc: func(options metaV1.ListOptions) (watch.Interface, error) {
+			return eventInterface.Watch(ctx, metaV1.ListOptions{})
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Log when pods are first added to the informer (at start-up).
+	syslog := logrus.WithFields(logrus.Fields{
+		"component": "event-informer",
+		"namespace": namespace,
+	})
+	for i := range events.Items {
+		syslog.Debugf("informer added event: %s", events.Items[i].Name)
+		cb(watch.Event{Object: &events.Items[i]})
+	}
+
+	return &informer{
+		cb:         cb,
+		name:       "event",
+		syslog:     syslog,
+		resultChan: rw.ResultChan(),
+	}, nil
+}
+
+func newNodeInformer(
+	ctx context.Context,
+	nodeInterface typedV1.NodeInterface,
+	cb informerCallback,
+) (*informer, error) {
+	nodes, err := nodeInterface.List(ctx, metaV1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	rw, err := watchtools.NewRetryWatcher(nodes.ResourceVersion, &cache.ListWatch{
+		WatchFunc: func(options metaV1.ListOptions) (watch.Interface, error) {
+			return nodeInterface.Watch(ctx, options)
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Log when nodes are first added to the informer (at start-up).
+	syslog := logrus.WithField("component", "nodeInformer")
+	for i := range nodes.Items {
+		syslog.Debugf("informer added node: %s", nodes.Items[i].Name)
+		cb(watch.Event{Object: &nodes.Items[i], Type: watch.Added})
+	}
 
-		i.syslog.Debugf("informer got new pod event for pod: %s %s", pod.Name, pod.Status.Phase)
-		i.cb(pod)
+	return &informer{
+		cb:         cb,
+		name:       "node",
+		syslog:     syslog,
+		resultChan: rw.ResultChan(),
+	}, nil
+}
+
+func (i *informer) run() {
+	ctx, cancel := context.WithCancel(context.TODO())
+	i.syslog.Debugf("%s informer is starting", i.name)
+	for {
+		select {
+		case event := <-i.resultChan:
+			if event.Type == watch.Error {
+				i.syslog.Warnf("%s informer emitted error %+v", i.name, event)
+				continue
+			}
+			i.cb(event)
+			cancel()
+		case <-ctx.Done():
+			panic(fmt.Sprintf("%s informer stopped unexpectedly: %s", i.name, ctx.Err()))
+		}
```

Comment:
Issue here -- placing the `cancel()` call causes informer to crash sooner than desired.
<img width="845" alt="Screen Shot 2023-07-25 at 3 14 30 PM" src="https://github.com/determined-ai/determined/assets/29692567/a43cc5f1-61bf-456e-adde-c3234a69f71f">

Reply:
reverted because informers crash with this implementation

Reply:
we shouldn't be calling cancel in the informer run loop.

i would've just made `run()` take a context like `run(ctx context.Context)`, pass `context.TODO()` for now in the pods actor, and then in tests, at the top of each test
```
ctx, cancel := context.WithCancel(context.Background())
defer cancel()
```
and use that context in the `run` calls in the test